package/MDAnalysis/auxiliary/EDR.py

```diff
@@ -0,0 +1,365 @@
+        The base AuxStep Class uses the time_selector 'Time' to return the
+        time value of each step."""
+        return self._select_data(key)
+
+    def _select_data(self, key: Union[str, None]) -> np.float64:
+        if key is None:
+            return
+        try:
+            return self._data[key]
+        except KeyError:
+            raise KeyError(f"'{key}' is not a key in the data_dict dictionary."
+                           " Check the EDRReader.terms attribute")
+
+
+class EDRReader(base.AuxReader):
+    """ Auxiliary reader to read data from a .edr file.
+
+    Default reader for .edr files. All data from the file will be read and
+    stored on initialisation.
+
+    Parameters
+    ----------
+    filename : str
+        Location of the file containing the auxiliary data.
+    **kwargs
+       Other AuxReader options.
+
+    See Also
+    --------
+    :class:`~MDAnalysis.auxiliary.base.AuxReader`
+    :func:`~MDAnalysis.coordinates.base.add_auxiliary`
+
+    Note
+    ----
+    The file is assumed to be of a size such that reading and storing the full
+    contents is practical.
+    """
+
+    format = "EDR"
+    _Auxstep = EDRStep
+
+    def __init__(self, filename: str, **kwargs):
+        self._auxdata = Path(filename).resolve()
+        self.data_dict = pyedr.edr_to_dict(filename)
+        self._n_steps = len(self.data_dict["Time"])
+        # attribute to communicate found energy terms to user
+        self.terms = list(self.data_dict.keys())
+        super(EDRReader, self).__init__(**kwargs)
+
+    def _read_next_step(self) -> EDRStep:
+        """Read next auxiliary step and update ``auxstep``.
+
+        Returns
+        -------
+        AuxStep object
+            Updated with the data for the new step.
+
+        Raises
+        ------
+        StopIteration
+            When end of auxiliary data set is reached.
+        """
+        auxstep = self.auxstep
+        new_step = self.step + 1
+        if new_step < self.n_steps:
+            auxstep._data = {term: self.data_dict[term][self.step + 1]
+                             for term in self.terms}
+            auxstep.step = new_step
+            return auxstep
+        else:
+            self.rewind()
+            raise StopIteration
+
+    def _go_to_step(self, i: int) -> EDRStep:
+        """ Move to and read i-th auxiliary step.
+
+        Parameters
+        ----------
+        i: int
+            Step number (0-indexed) to move to
+
+        Returns
+        -------
+        :class:`EDRStep`
+
+        Raises
+        ------
+        ValueError
+            If step index not in valid range.
+        """
+        if i >= self.n_steps or i < 0:
+            raise ValueError("Step index {0} is not valid for auxiliary "
+                             "(num. steps {1})".format(i, self.n_steps))
+        self.auxstep.step = i - 1
+        self.next()
+        return self.auxstep
+
+    def read_all_times(self) -> np.ndarray:
```

Comment:
I could go either way here - Of course other units of time would have to be converted to make associating with the correct trajectory timesteps possible, but because we cannot really predict what sorts of units the other data in other AuxReaders might have, it might be better to warn and/or fail cleanly so users think about their units carefully and change them themselves were appropriate?

Reply:
@fiona-naughton was this discussed with the XVG reader?

I would definitely advocate for converting unit "coords-style" or make it very clear (and say it loudly) that the units are not converted and are in the native units as read from file.

Otherwise we're going to get folks comparing Angstroms to nanometers and get confused real quick.

Reply:
I think for the formats where the units are known eg. GROMACS ones  and other well defined simulation engines, we should enforce convert_units as in `coords`.

This is done by adding a `units` class attribute and then enforcing the conversion with a base class method `convert_X_to_native. ` See [here](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/base.py#L356) for an example.

Reply:
There is an inconvenience associated with this: While some units may be easily guessed from the dictionary keys (`Time` in an EDR file would be in ps, for example, and energies in kJ/mol), this is less obvious for some of the more "exotic" possible entries in aux files like `#Surf*SurfTen`, `Lamb-Protein`, or `Pres. DC`

In the EDR file I've created for testing purposes, the following units appear:
  - kJ/mol
  - K
  - bar
  - unitless
  - nm
  - nm ^ 3
  - kg / m^3 
  - bar * nm
  - nm / ps
 
Unfortunately, this unit information is no longer present at the level of the EDRReader. It is read from the file in pyedr, but not actually part of the return value. So we would need some way to reliably guess that the unit for "`Surf*SurfTen` is bar * nm, so we could then convert bar into the equivalent MDA base unit of, I guess, some factor times kJ/(mol·Å) / Å^2? 
We can't only change the easier units like time or velocities, because the data would then no longer be consistent, and the only reliable way I can think of to obtain the units is to read them from the file - guessing seems inappropriate here. 

So the two options I believe we have are not to change any units, or to rewrite pyedr and EDRReader so unit information is retained and then available for conversion. Between these options, I would favour the former (and warn very loudly that no units are changed by us). 

What do you think? Am I maybe missing a third option?

Reply:
I disagree, units are extremely important, not being able to expose to folks what their units are is just dangerous.

My view is that we should expose the units in pyedr even if that means making `pint` a requirement for pyedr.

Reply:
I also disagree, inconsistent units is just too confusing. 

I would be happy with a `units` class attribute hardcode for now if we want to push ahead  with merging this PR ASAP, but I do think we need to do the conversion. EDR files always come from GROMACS which always uses the same [units](https://manual.gromacs.org/documentation/2019/reference-manual/definitions.html) so I think the hardcode is safe enough. We could then push `pint` integration down the track. What do you think @IAlibay?

 Nevertheless this brings up an important point. I will raise an issue for Aux units for further general discussion.

Reply:
For right now, I'd hard-code known units. That's consistent with the trajectory I/O.

Can we have a dict of `auxterm => unit` associated with the EDRReader? (Where `auxterm` is whatever the user ultimately uses to access the data and not what it's called in the EDR file.)

If we were to go to a more integrated unit representation then we would change it.

Reply:
We can move long term discussion to  #3792 :)

Reply:
Just a word of caution here. I have seen cases where gromacs was used with exotic units. If all your input files are consistent gromacs does not care if your simulation uses reduced units or something. Then it might be non-sensical to convert the units in mdanalysis.

It is a rare use case, though. At least I think it is. But it would be good to have a way to have the raw values from the file somehow; even if it implies using the correct unspoken incantation.

Reply:
I don't think #3792 is the right place to have this discussion for EDR files. The dealing of interpreting the right units from an EDR file is a pyedr problem not an MDAnalysis one. I honestly don't think we should have a dictionary with just a list of arbitrary fields and their expected units in MDA. If you want that, this should be in pyedr.

Conversion to MDA internals then happen on our end if necessary.

Reply:
I’m fine with having pyedr having to deal with units. In principle, we can decide to just pass everything through as is and the user needs to know the units. (The same applies to XVG anyway and will apply to numpy and CSV). The time, however, is important and we somehow need to en able to get it in the MDA unit of ps.

Reply:
At the very least, pyedr could expose the units. It is the one supposed to know them. Converting, though, should probably be done in mdanalysis since it is the one knowing about the targets.

I am all about using a unit framework, but then we need to use it throughout the library, not just aux.

Reply:
Yeah what I'm trying to say is pretty much on the same line of thought as @jbarnoud - we shouldn't be guessing the units here, pyedr should be exposing them, that will be significantly cleaner and faster to implement / update / maintain.

We then convert them as we do coordinate readers here in the MDA library.

If we make pint a core requirement of pyedr then we can just use it here for conversion, that being said I'd also live with just a conversion table as per coordinate readers for now.

Reply:
Hard-coding the units would not be straightforward, because what is written to EDR files depends on the input simulation parameters. Between the cat.edr test file for pyedr and my minimal test file for the EDRReader, for instance, there are 40-ish terms that appear in one but not the other. 

So I agree that for unit handling, pyedr needs to pass the units on. These could be returned as a second dictionary of entry name -> unit for now and saved at the level of the individual EDRReader.

Reply:
@hmacdope @IAlibay @orbeckst @jbarnoud I have now opened a [PR](https://github.com/MDAnalysis/panedr/pull/56) that makes pyedr also return the units as read from the EDR file. If we get that merged and released, I can then change the EDRReader to be aware of the units as well.

Reply:
I am closing this thread as resolved, now that we have units coming from pyedr. Please reopen if there's still something that needs addressing.